Bump shellcheck to v0.7.0

odinuge opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Changelog:
https://github.com/koalaman/shellcheck/blob/master/CHANGELOG.md

We have some new erros due to the version bump.

- SC2034: VARIABLE_XYZ appears unused. Verify use (or export if used externally).
    - Applies to all scripts we source in other scripts
- SC2039: In POSIX sh, set option pipefail is undefined.
    - Applies to files using it with "sh" instead of "bash" in the shebang
- SC2054: Use spaces, not commas, to separate array elements.
    - Fixing Should make no difference in the code
- SC2128: Expanding an array without an index only gives the first element.
    - Fixing Should make no difference in the code
- SC2251: This ! is not on a condition and skips errexit. Use `&& exit 1` instead, or make sure $? is checked.
    - Not 100% sure if we can swap to `&& exit 1`. Applies to a lot of
    test code.

All changes should be straight forward to fix, but will be done in a
separate PR.

Errors before updating `hack/.shellcheck_failures`:  https://gist.github.com/odinuge/7e49124448e7dd0cf1d758e2af294a9b

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes # `N/A`

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

odinuge committed:
Bump shellcheck to v0.7.0

Changelog:
https://github.com/koalaman/shellcheck/blob/master/CHANGELOG.md

We have some new erros due to the version bump.

- SC2034: VARIABLE_XYZ appears unused. Verify use (or export if used externally).
    - Applies to all scripts we source in other scripts
- SC2039: In POSIX sh, set option pipefail is undefined.
    - Applies to files using it with "sh" instead of "bash" in the shebang
- SC2054: Use spaces, not commas, to separate array elements.
    - Fixing Should make no difference in the code
- SC2128: Expanding an array without an index only gives the first element.
    - Fixing Should make no difference in the code
- SC2251: This ! is not on a condition and skips errexit. Use `&& exit 1` instead, or make sure $? is checked.
    - Not 100% sure if we can swap to `&& exit 1`. Applies to a lot of
    test code.

All changes should be straight forward to fix, but will be done in a
separate PR.

odinuge commented:
/cc @kubernetes/bash-firefighters

BenTheElder commented:
😞

BenTheElder commented:
> All changes should be straight forward to fix, but will be done in a
separate PR.

if the fixes are trivial, can we just fix them now? when we update eg go we usually fix lint errors then.

odinuge commented:
/test pull-kubernetes-e2e-gce-device-plugin-gpu

odinuge commented:
> if the fixes are trivial, can we just fix them now? when we update eg go we usually fix lint errors then.

If that is preferred, I can do so. Not 100% sure about  [SC2251](https://github.com/koalaman/shellcheck/wiki/SC2251) tho, so I will take a look

odinuge committed:
Fix shellcheck failures SC2034

fejta commented on hack/.shellcheck_failures:
```diff
@@ -10,3 +10,12 @@
 ./cluster/gce/util.sh
 ./cluster/log-dump/log-dump.sh
 ./cluster/pre-existing/util.sh
+./test/cmd/apply.sh
```
Why are we adding things? Can we first merge PRs to fix these issues? Ideally we get the whitelist down to zero.

odinuge replied:
Yup! Working on it now ref the comment from @BenTheElder ! So ill add the fixes into this PR.

Just doing some investigation on https://github.com/koalaman/shellcheck/wiki/SC2251, that may be masking some test failures. Updating the code as we speak.

We have a bunch of test code like this:

```bash
set -o errexit
[...]
! command_that_should_fail
[...]
```

This test will succeed no matter what the result of `command_that_should_fail` is.

odinuge replied:
Or do you want me to create separate PRs for each fix?

fejta replied:
Either way. In this PR is great too.

odinuge replied:
Updated this PR now, but splitted different stuff into different commits. :smile:

fejta reviewed (commented):
Thanks!
/approve

odinuge committed:
Fix shellcheck failures SC2128

odinuge committed:
Fix shellcheck failures SC2251

https://github.com/koalaman/shellcheck/wiki/SC2251

This may be masking some test failures.

We have a bunch of test code like this:

set -o errexit
[...]
! kubectl get pod wrong-pod
[...]

This test will succeed no matter what the result (return code) of kubectl is.

odinuge commented on cluster/gce/gci/mounter/stage-upload.sh:
```diff
@@ -20,7 +20,6 @@
 # put in configure.sh, which is uploaded via GCE custom metadata.
 
 set -o errexit
-set -o pipefail
```
https://github.com/koalaman/shellcheck/wiki/SC2039

We can switch the shebang to bash instead of sh instead of removing this line (I think I am in favor of that). The image used include bash as well.

BenTheElder replied:
+1, let's switch to bash.

odinuge commented on hack/lib/golang.sh:
```diff
@@ -1,4 +1,5 @@
 #!/usr/bin/env bash
+# shellcheck disable=SC2034 # Variables sourced in other scripts.
```
When shellcheck disable statements are placed here they apply to the whole file (since this applies to several lines)

BenTheElder replied:
on the fence about this, we should probably be verifying which we export and which are unused and remove things that are unused.

odinuge committed:
Fix failing service test

fejta commented:
/retest

fejta reviewed (approved):
/lgtm

fejta commented:
/retest

These failures might be related to this PR

odinuge commented:
Ref. comment from @BenTheElder about `SC2034`, should I sort that out in this PR or should I create a new PR fixing it?

The PR is only missing approval from `cluster/gce/gci/OWNERS`, but the two people in that owners file look quite inactive. Guess I can move that change into another PR.

/assign dchen1107 yguo0905

fejta commented:
/assign @bowei @cjcullen mtaufen

for gce approval

SataQiu reviewed (approved):
Great!
/lgtm
/approve

BenTheElder commented:
yes I think this is OK, I'm not a huge fan of globally disabling lints in files but in this particular case it seems reasonable enough.

thanks for working on this

BenTheElder commented on test/images/pets/redis-installer/on-start.sh:
```diff
@@ -25,11 +25,11 @@ PORT=6379
-    if [[ "${LINE}" == *"${HOSTNAME}"* ]]; then
-        sed -i -e "s|^bind.*$|bind ${LINE}|" ${CFG}
-    elif [ "$(/opt/redis/redis-cli -h "${LINE}" info | grep role | sed 's,\r$,,')" = "role:master" ]; then
+    if [[ "${LINE[0]}" == *"${HOSTNAME}"* ]]; then
```
is this commit identical in behavior?

odinuge replied:
Yes, it is!

Since we use `-a` in read, the result (`LINE`) is an array of the words on the given line.

When `LINE` is an array, `${LINE}` is equivalent to `${LINE[0]}`, and results in the first element in the array.

More info can be found here: https://github.com/koalaman/shellcheck/wiki/SC2128

mtaufen commented:
/approve
but
/hold
for Ben's latest comment (https://github.com/kubernetes/kubernetes/pull/84249#pullrequestreview-309515954)

mtaufen commented:
/priority important-longterm

odinuge commented:
Ref. https://github.com/kubernetes/kubernetes/pull/84249#discussion_r340991817

/hold cancel

odinuge commented:
/test pull-kubernetes-e2e-gce
